tools/codegen/api/translate.py

```diff
@@ -92,9 +93,34 @@ def translate(
 
         # While we're at it, do some simple forward inference, looking through
         # constructors.
+        #
+        # NB: When should you do forward inference versus backward inference?
+        # The general idea:
+        #
+        #   - Backward inference WHEN the goal gets smaller
+        #   - Forward inference WHEN the hypothesis gets smaller
```

Comment:
Ok, I need to read the pdf that you linked in more detail. But how do you picture this evolving over time? My understanding is:
- right now, `translate()` does simple, "single-step" forward inference"
- This PR increases the context size that `translate()` works with by a bit (but so do lots of other codegen related PR's that introduce new C++ expressions)
- It's possible that as we lean on `translate()` more heavily, we could introduce cases where it gets really slow, or doesn't even terminate
- we could alleviate those problems with some of the ideas you put here, like automatically figuring out when it would be less compute-intensive to run forward vs backward inference, memoizing expressions, etc